[Async Refactoring] Support Obj-C style bool flag checks

hamishknight opened the pull request:
Add an additional case to `CallbackCondition` to support boolean checks, and add the necessary classification logic to determine whether a bool flag check is for a success or failure path.

The logic currently first checks to see if the async alternative has a convention that specifies which parameter the flag is, and whether it's a success or failure flag. If so, it classifies the flag check accordingly. If there's no async convention specified, we use a heuristic to see if the error parameter is force unwrapped in the failure block. If so, we treat that as the error path. If there's no force unwrap of the error, we leave the condition unhandled.

rdar://74063899

hamishknight commented:
@swift-ci please test

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -4841,12 +4830,108 @@ struct CallbackClassifier {
+    case ConditionType::SUCCESS_PATTERN:
+      break;
+    }
+    return ClassifiedCondition(Cond, Path);
```
Love the double flip 👍

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -4830,10 +4900,66 @@ struct CallbackClassifier {
+
+      std::pair<bool, Stmt *> walkToStmtPre(Stmt *S) override {
+        // Don't walk into new explicit scopes, we only want to consider force
+        // unwraps in the immedidate conditional body.
```
Typo (should be "immediate")

bnbarham replied:
Also, does this need to be limited to the immediate body?

hamishknight replied:
Also replied to this below, but the idea of restricting it to the immediate body is so we don't accidentally consider bits of code that take place in nested scopes that may have additional information about whether they can force unwrap the error.

bnbarham commented on test/refactoring/ConvertAsync/convert_bool.swift:
```diff
@@ -0,0 +1,271 @@
+
+// All 4 of the below should generate the same refactoring.
+
+// RUN: %refactor -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):1 -I %S/Inputs -sdk %clang-importer-sdk-path  | %FileCheck -check-prefix=BOOL-WITH-ERR %s
```
These don't need -I or the SDK path right?

hamishknight replied:
They're needed to resolve the imports in the file, though it seems `swift-refactor` will still return 0 in that case, still I think it it's probably best to include them so the errors don't clutter the stderr when debugging a failure, what do you think?

bnbarham replied:
Ah true, yep - keeping them sounds good.

bnbarham commented on test/refactoring/ConvertAsync/convert_bool.swift:
```diff
@@ -0,0 +1,271 @@
+
+// RUN: %refactor -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):1 -I %S/Inputs -sdk %clang-importer-sdk-path  | %FileCheck -check-prefix=BOOL-WITH-ERR2 %s
+boolWithErr { success, err in
+  if success == true && err == nil {
```
May as well check for just `if success` as well

bnbarham commented on test/refactoring/ConvertAsync/convert_bool.swift:
```diff
@@ -0,0 +1,271 @@
+// RUN: %refactor -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):1 -I %S/Inputs -sdk %clang-importer-sdk-path  | %FileCheck -check-prefix=BOOL-DONT-HANDLE2 %s
+boolWithErr { success, err in
+  if !success {
+    func doThings() {
```
See comment above - is there a good reason not to handle this case? Ie. if the force unwrap is used at all in the scope (nested or not), treat it as the error condition?

hamishknight replied:
The problem is that would count for things like this:

```
if !success {
  func doThings() {
    print(err!)
  }
  if err != nil {
    doThings()
  }
}
```

Where the `!success` condition doesn't appear to be relevant to the force unwrap. We could probably perform a more in-depth heuristic that also looks for other references to `err`, but I'm not sure whether or not that's worth it. What do you think?

bnbarham replied:
Makes sense, I wouldn't expect that case to be super common though. Another would be returning multiples bools and checking each both (nested). Though again, I assume that's uncommon. Erring on the side of leaving it and adding placeholders is probably okay for now though 👍

bnbarham commented on test/refactoring/ConvertAsync/convert_bool.swift:
```diff
@@ -0,0 +1,271 @@
+
+// RUN: %refactor -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):1 -I %S/Inputs -sdk %clang-importer-sdk-path  | %FileCheck -check-prefix=BOOL-WITH-ERR-SILLY %s
+boolWithErr { success, err in
+  if success == false && err == nil {
```
I agree kind of silly, but this sort of implies that `success` is actually `error`. We *could* handle it as such... Thoughts?

hamishknight replied:
Yeah that's a fair point! It seems we should really make the determination of whether the flag is a success check or failure check from the branch in which the force unwrap occurs rather than the other way around. I think it would then also make sense to get rid of the bit of logic that discounts a flag condition if its path disagrees with the rest of the condition paths – I'll go ahead and make both those changes.

hamishknight replied:
Hmm, yeah we could handle that – it seems we'd want to change the condition path of the flag check if it disagrees with the rest of the conditions rather than dropping it entirely. I'll make that change.

On a related note, it also seems we should initially make the determination of whether the flag is a success check or failure check based on the branch in which the force unwrap occurs rather than the other way around, which would let us classify a flag as an `error` check if we then force unwrap the error.

hamishknight replied:
Hmm, the problem is that we're only classifying it as a success flag to begin with because of the force unwrap of the error in the body, so even if we classify both checks as success paths, the code still doesn't make much sense. We could apply a more general heuristic that allows for bool condition checks if they're also with other conditions that are able to classify as success or failure, but I don't really have a good sense of how often that comes up.

On a related note, I've just noticed that if the force unwrap of the error in this example was in an `else` block here, we would still classify the condition as the same, which doesn't seem right. It seems we should adjust the classification logic so we make the determination of whether the flag is a success check or failure check based on the branch in which the force unwrap of the error occurs rather than the other way around, which would let us classify a flag as an `error` check if we then force unwrap the error.

bnbarham replied:
Probably a couple of problems here really. Yes, ideally if the force unwrap was in the else we would then consider `success` to be a flag. Could get a bit tricky when the `else` can itself be another `if` though.

Another problem is that if we don't classify the `success` as being a flag then there are unhandled conditions and so we don't remove the `if` at all. Ideally we'd classify the whole if as being in the success block and then remove the `&& err == nil`. But that's an existing issue so... 🤷

bnbarham commented on test/refactoring/ConvertAsync/convert_bool.swift:
```diff
@@ -0,0 +1,271 @@
+
+func boolWithErr(completion: (Bool, Error?) -> Void) {}
+func multipleBoolWithErr(completion: (String?, Bool, Bool, Error?) -> Void) {}
+func optionalBoolWithErr(completion: (String?, Bool?, Bool, Error?) -> Void) {}
```
Isn't actually a test for this one

hamishknight replied:
Ah, good spot!

hamishknight replied:
Turns out there was a bit of logic missing needed to handle this case! Updated now.

bnbarham reviewed (commented):
Thanks for the cleanup as usual 🙇. No major comments

hamishknight commented:
Updated to handle the unwrap case, and changed the flag classification logic to consider the branch the force unwrap happens in, rather than assuming the flag always denotes success. Also noticed we were missing a couple of `getSemanticsProvidingExpr` calls, added those in.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Linux

bnbarham commented on test/refactoring/ConvertAsync/convert_bool.swift:
```diff
@@ -0,0 +1,382 @@
+  if failure && err == nil {
+    print("wat")
+  }
+  if failure && err != nil {
```
Do we handle `if failure, let err = err`?

hamishknight replied:
Yes, let me add a test for that along with an `else if` test

hamishknight replied:
Done, how does this look?

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Linux

hamishknight commented:
@swift-ci please clean test Linux

bnbarham commented on test/refactoring/ConvertAsync/convert_bool.swift:
```diff
@@ -0,0 +1,448 @@
+  }
+}
+
+// It's a little unfortunate that print("a \(<#err#>!)") gets classified in the success
+// block below, but it doesn't seem like a case that's likely to come up, as optBool
+// would then be inaccessible in the error block.
```
For both `a` and `f` we could just assume that if there's a force unwrap, that == the error branch (only where the conditions aren't explicitly success/error and involve known params). I'm happy with this as is though, just thought I'd mention it.

hamishknight replied:
Yeah it's possible we'll want to generalise the heuristic to better handle cases like these, though I'm not really sure how often they'll come up in practice as it seems a little odd to e.g check a success param for non-nil and then assume you're on an error path without looking at the success param again. I think for now at least I'd rather tend on the side on keeping the heuristic a little more limited and seeing what kind of feedback we get from users.

bnbarham reviewed (approved):
One final comment, but we can do that in another PR if we want it.

hamishknight committed:
[Refactoring] Rework how conditions are classified

- Expand ConditionType to include `.success` and
`.failure` patterns.
- Introduce ConditionPath to represent whether a
classified condition represents a success or
failure path.
- Add methods to CallbackClassifier that deal with
deciding whether a given condition is a success or
failure path.
- Use these methods to simplify `classifyConditional`.

hamishknight committed:
[Refactoring] Remove ConditionType::INVALID

This will make it a little nicer to switch on
ConditionType.

hamishknight committed:
[Refactoring] Add some calls to getSemanticsProvidingExpr

hamishknight committed:
[Refactoring] Support Obj-C style bool flag checks

Add an additional case to `CallbackCondition` to
support boolean checks, and add the necessary
classification logic to determine whether a bool
flag check is for a success or failure path.

The logic currently first checks to see if the
async alternative has a convention that specifies
which parameter the flag is, and whether it's a
success or failure flag. If so, it classifies the
flag check accordingly. If there's no async
convention specified, we use a heuristic to see if
the error parameter is force unwrapped in the
failure block. If so, we treat that as the error
path. If there's no force unwrap of the error, we
leave the condition unhandled.

rdar://74063899

hamishknight committed:
[swift-refactor] Support -I and -sdk

hamishknight commented:
@swift-ci please smoke test

hamishknight commented:
@swift-ci please test Windows
